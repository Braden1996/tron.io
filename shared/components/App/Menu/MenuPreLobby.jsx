import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import Redirect from 'react-router-dom/Redirect';

import {
  lobbySetName
} from "../../../state/lobby/actions";


class MenuPreLobby extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      connectRandomLobby: null
    };

    this.onNameChange = this.onNameChange.bind(this);
    this.createLobby = this.createLobby.bind(this);
  }

  onNameChange(event) {
    this.props.setName(event.target.value);
  }

  createLobby(event) {
    this.setState({connectRandomLobby: "RandomLobbyString"});
  }

  render() {
    return (
      <div>
        {this.state.connectRandomLobby !== null &&
          <Redirect to={{pathname: this.state.connectRandomLobby}} push />
        }

        <h2>Welcome to Tron!</h2>
        <p>If you're looking to start a new game, just append a unique server name to the url - or click the button below!</p>
        <br />
        Name:
        <input type="text" name="menu_myname" onChange={this.onNameChange} value={this.state.name} /><br />
        <button onClick={this.createLobby}>Create Lobby</button><br />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    name: state.get('lobby').get('myName'),
  };
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    setName: lobbySetName,
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(MenuPreLobby);
