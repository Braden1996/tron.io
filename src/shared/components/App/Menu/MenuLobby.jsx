import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { Link } from 'react-router';

import {
  updateLobbyConnect
} from "../../../actions/gamelobby.js";


class MenuLobby extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    console.log(`Component mount init Lobby connect: ${this.props.lobbyKey}`);
    this.props.updateLobbyConnect(this.props.lobbyKey);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.lobbyKey !== this.props.lobbyKey) {
      console.log(`Component lobby key changed: ${nextProps.lobbyKey}`);
      this.props.updateLobbyConnect(nextProps.lobbyKey);
    }
  }

  render() {
    const plyList = this.props.players.map((ply, idx) =>
      <li key={ply.get("id")}>
        {ply.get("name")}
      </li>
    );

    let beginButton;
    if (this.props.started) {
      beginButton = (
        <button onClick={this.props.onEndGame}>
          {"End"} Game
        </button>
      )
    } else {
      beginButton = (
        <button onClick={this.props.onStartGame}>
          {"Begin"} Game
        </button>
      )
    }

    return (
      <div>
        <h2>{this.props.lobbyKey}</h2>
        <figure>
          <figcaption>Players:</figcaption>
          <ol>{plyList}</ol>
          <p>To invite a friend, send them a copy of your url.</p>
        </figure>
        {!this.props.started && this.props.players.size < 16 &&
          <button onClick={this.props.onAddComputer}>Add computer player</button>
        }
        {beginButton}
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    players: state.players
  };
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    updateLobbyConnect: updateLobbyConnect
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(MenuLobby);
