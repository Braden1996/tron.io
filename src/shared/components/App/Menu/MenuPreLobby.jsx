import React from "react";
import { Redirect } from 'react-router';


export default class MenuPreLobby extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "Player",
      connectRandomLobby: null
    };

    this.onNameChange = this.onNameChange.bind(this);
    this.createLobby = this.createLobby.bind(this);
  }

  onNameChange(event) {
    this.setState({name: event.target.value});
  }

  createLobby(event) {
    this.setState({connectRandomLobby: "RandomLobbyString"});
  }

  render() {
    return (
      <div>
        {this.state.connectRandomLobby !== null &&
          <Redirect to={{pathname: this.state.connectRandomLobby}} push />
        }

        <h2>Welcome to Tron!</h2>
        <p>If you're looking to start a new game, just append a unique server name to the url - or click the button below!</p>
        <br />
        Name:
        <input type="text" name="menu_myname" onChange={this.onNameChange} value={this.state.name} /><br />
        <button onClick={this.createLobby}>Create Lobby</button><br />
      </div>
    )
  }
}
