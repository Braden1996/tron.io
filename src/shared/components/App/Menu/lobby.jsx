import React from "react";


export default class MenuLobby extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    const plyList = this.props.players.map((ply, idx) =>
      <li key={ply.get("id")}>
        {ply.get("name")}
      </li>
    );

    let beginButton;
    if (this.props.started) {
      beginButton = (
        <button onClick={this.props.onEndGame}>
          {"End"} Game
        </button>
      )
    } else {
      beginButton = (
        <button onClick={this.props.onStartGame}>
          {"Begin"} Game
        </button>
      )
    }

    return (
      <div>
        Lobby code: <input type="text" name="menu_lobby_code" value={this.props.code} disabled />
        <figure>
          <figcaption>Players:</figcaption>
          <ol>{plyList}</ol>
        </figure>
        {this.props.players.size < 16 &&
          <button onClick={this.props.onAddComputer}>Add computer player</button>
        }
        {beginButton}
      </div>
    )
  }
}
