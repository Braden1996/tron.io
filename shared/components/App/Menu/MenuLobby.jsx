import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import Link from 'react-router-dom/Link';

import { lobbyConnect } from "../../../state/lobby/actions";
import { addComputer } from "../../../state/input/actions";


class MenuLobby extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const lobbyKey = this.props.match.params.lobbykey;
    this.props.lobbyConnect(lobbyKey, 'Player Name', '#0f0');
  }

  componentWillReceiveProps(nextProps) {
    // In the case where this component is mounted before our socket saga is
    // ready, our lobbyConnect dispatch will not be received.
    // We thus check to see if the socket saga 'ready' state has been enabled,
    // then dispatch a new lobbyConnect dispatch.
    const tryLobbyConnectAgain = !this.props.ready && nextProps.ready;

    const lobbyKey = this.props.match.params.lobbykey;
    const nextLobbyKey = nextProps.match.params.lobbykey;
    if (tryLobbyConnectAgain || lobbyKey !== nextLobbyKey) {
      this.props.lobbyConnect(nextLobbyKey, 'Player Name', '#0f0');
    }
  }

  render() {
    const plyList = this.props.players.map((ply, idx) =>
      <li key={ply.id}>
        {ply.name}
      </li>
    );

    let beginButton;
    if (this.props.started) {
      beginButton = (
        <button onClick={this.props.onEndGame}>
          {"End"} Game
        </button>
      )
    } else {
      beginButton = (
        <button onClick={this.props.onStartGame}>
          {"Begin"} Game
        </button>
      )
    }

    return (
      <div>
        <h2>{this.props.lobbyKey}</h2>
        { this.props.lobbyConnected ? "Connected" : "Connecting" }
        <figure>
          <figcaption>Players:</figcaption>
          <ol>{plyList}</ol>
          <p>To invite a friend, send them a copy of your url.</p>
        </figure>
        {!this.props.started && this.props.players.length < 16 &&
          <button onClick={this.props.addComputer}>Add computer player</button>
        }
        {beginButton}
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const sockets = state.get('sockets');
  const gameState = state.get('lobby').get('gameState');
  return {
    players: gameState ? gameState.players : [],
    lobbyConnected: state.get('lobby').get('connected'),
    ready: sockets && sockets.get('receiveReady') && sockets.get('sendReady'),
  };
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    lobbyConnect: lobbyConnect,
    addComputer: addComputer,
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(MenuLobby);
