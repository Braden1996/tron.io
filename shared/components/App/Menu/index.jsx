import React from 'react';
import Switch from 'react-router-dom/Switch';
import Route from 'react-router-dom/Route';

import MenuLobby from './MenuLobby';
import MenuPreLobby from './MenuPreLobby';

class Menu extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div>
        <h1>Lobby</h1>
        {/* I cannot figure out how to use a JS condition inside a Switch. */}
        <Switch>
          <Route exact path="/:lobbykey" component={MenuLobby} />
          <Route path="/" component={MenuPreLobby}/>
        </Switch>
      </div>
    );
  }
}

export default Menu;
