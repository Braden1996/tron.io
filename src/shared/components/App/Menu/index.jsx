import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import MenuLobby from "./lobby.jsx";
import MenuPreLobby from "./prelobby.jsx";

import {
  updateStartGame,
  updateFinishGame,
  updateSpeed
} from "../../../game/state/actions/game.js";
import {
  addPlayer,
  resetPlayers
} from "../../../game/state/actions/players.js";


class Menu extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      computerPlayers: 0
    }
  }

  onStartGame() {
    this.props.startGame(true);
    this.props.finishGame(false);
  }

  onEndGame() {
    this.props.startGame(false);
    this.props.resetPlayers();
  }

  onAddComputer() {
    const computerPlayers = this.state.computerPlayers + 1;
    this.props.addPlayer(`Computer ${computerPlayers}`);
    this.setState({computerPlayers});
  }

  onCreateLobby(speed, name) {
    this.props.updateSpeed(speed);
    this.props.addPlayer(name);
  }

  render() {
    let lobby = null;
    if (this.props.players.size > 0) {
      lobby = <MenuLobby
        started={this.props.started}
        players={this.props.players}
        code={"blahblah264276"}
        onStartGame={this.onStartGame.bind(this)}
        onEndGame={this.onEndGame.bind(this)}
        onAddComputer={this.onAddComputer.bind(this)}
      />;
    } else {
      lobby = <MenuPreLobby
        speed={this.props.speed}
        onCreateLobby={this.onCreateLobby.bind(this)}
      />;
    }

    return (
      <div>
        <h1>Lobby</h1>
        {lobby}
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    started: state.game.get("started"),
    finished: state.game.get("finished"),
    arenaSize: state.game.get("arenaSize"),
    playerSize: state.game.get("playerSize"),
    speed: state.game.get("speed"),
    players: state.players
  };
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    startGame: updateStartGame,
    finishGame: updateFinishGame,
    updateSpeed: updateSpeed,
    addPlayer: addPlayer,
    resetPlayers: resetPlayers
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Menu);
