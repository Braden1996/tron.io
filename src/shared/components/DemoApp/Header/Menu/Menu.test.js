/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';
import Menu from './Menu';

describe('<Home />', () => {
  test('renders', () => {
    const wrapper = shallow(<Menu />);
    expect(wrapper).toMatchSnapshot();
  });
});
