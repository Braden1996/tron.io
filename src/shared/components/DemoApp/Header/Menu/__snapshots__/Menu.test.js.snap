exports[`<Home /> renders 1`] = `
<ul
  style={
    Object {
      "marginTop": "1rem",
    }
  }>
  <li>
    <Link
      activeClassName=""
      activeOnlyWhenExact={false}
      activeStyle={Object {}}
      className=""
      isActive={[Function]}
      replace={false}
      style={Object {}}
      to="/">
      Home
    </Link>
  </li>
  <li>
    <Link
      activeClassName=""
      activeOnlyWhenExact={false}
      activeStyle={Object {}}
      className=""
      isActive={[Function]}
      replace={false}
      style={Object {}}
      to="/about">
      About
    </Link>
  </li>
</ul>
`;
